### PR TITLE
Add an array API to scipy.sparse

### LINKED ISSUE #7462: numpy.power on sparse matrix attempts to do matrix power not elementwise
numpy power seems to go to a matrix power routine for sparse matrices.

### Reproducing code example:
```
    import scipy.sparse as sps
    import numpy as np
    a = np.array([[1,2,3],[4,0,0]])

    np.power(a,2)

    np.power(sps.csr_matrix(a),2) # this fails 

    sps.csr_matrix(a).power(2).todense()

```

### Error message:
    Traceback (most recent call last):

      File "<ipython-input-7-513a059a6886>", line 1, in <module>
      np.power(sps.csr_matrix(a),2)

      File "/home/sean/anaconda2/lib/python2.7/site-packages/scipy/sparse/base.py", line 539, in __pow__
    raise TypeError('matrix is not square')

    TypeError: matrix is not square

### Scipy/Numpy/Python version information:
```
('0.19.0', '1.12.1', sys.version_info(major=2, minor=7, micro=13, releaselevel='final', serial=0))
```

### LINKED ISSUE #8162: Sparse ndarray class
While sitting with @stefanv last week we discussed (and then @stefanv started implementing) sparse array classes that use the same code as the existing matrix classes, but instead satisfy the ndarray interface (only up to two dimensions).  For example slicing out rows would result in 1d arrays rather than 2d matrices, the `*` operator means elementwise multiply rather than matrix multiply, etc..

This would be valuable for a number of communities and, in particular, anyone trying to use sparse arrays and numpy arrays interchangably, as happens in Pandas, Scikit-Learn, and Dask.array.

I thought I'd raise this issue to track this work (and also to subtly prod @stefanv).

Also cc @NelleV and @njsmith

### LINKED ISSUE #14497: MAINT: Sparse matrices: switch the API to produce arrays and not np.matrix
**Is your feature request related to a problem? Please describe.**

This is a continuation of the reasoning of #14494. Many sparse matrix methods return a `numpy.matrix`. An offender that I found includes `.sum(axis=i)`, but there are likely more.

**Describe the solution you'd like**

I believe that when reasonable these methods should return a 1D numpy array, however a 2D array with shape 1×N or N×1 for backwards compatibility is also an option.

What I am not quite sure about is whether this should follow a deprecation cycle, and if yes, what should be the course of action.

### PR DESCRIPTION
**Pull request by stefanv**
This PR adds matching array classes for the bsr, coo, csc, csr, dia, dok, and lil matrices,
as a thin layer on top of the existing sparse matrix implementation.

For each of the new sparse array classes, we restore element-wise
multiplication.  In the existing implementation, we switch to using the `@` operator
(instead of `*`, which behaves differently for arrays and matrices).

The existing matrix implementation often explicitly creates matrices
or sparse matrices.  To address that, we attach to each sparse array /
matrix various container attributes (`_coo_container`,
`_lil_container`, etc.).  These map to `_coo_matrix` and `_lil_matrix`
for sparse matrices, and to `_coo_array` and `_lil_array` for sparse
arrays.  Thus, the implementation can be agnostic of the container type.

For the same reason `_ascontainer` and `_container` methods are
introduced, which map to `asmatrix`/`matrix` and `asarray`/`array`
respectively.

---

One implementation detail I would like to improve is the detection of sparse arrays and matrices.  For now, that is done using `if "matrix" in self.__class__.__name__`, but that could be a problem if the user subclasses from these.

The test suite contains the failure modes I anticipated, but can be expanded to provide more thorough coverage of all basic operations.

---

Closes #7462
Closes #8162
Closes #14497

### CODE REVIEW

**Comment by stefanv** [comment]
@ev-br I agree that we want to replicate the ndarray interface as closely as possible.  I could not see an easy way to making sparse arrays support 1D (the 2D shape is deeply embedded in the code), so instead I modified it so that 1D slice results are dense ndarrays.  This is not optimal, but it does result in consistent array semantics; and hopefully having to deal with non-sparse row or column slices will not typically be too expensive.

**Comment by rgommers** [comment]
> Added `__array__` conversion as requested in #4239

That's probably not a good idea - there are `asarray` calls all over the place, and densifying a sparse matrix/array still should be an explicit operation.

**Comment by stefanv** [comment]
> > Added `__array__` conversion as requested in #4239
> 
> That's probably not a good idea - there are `asarray` calls all over the place, and densifying a sparse matrix/array still should be an explicit operation.

OK, I reverted that change for now.

**Comment by ev-br** [comment]
> @ev-br You mean for the .A attribute, correct? I think the slicing is still something we should consider separately (since we have no notion of a 1D sparse container right now). But we could implement that too.

Indeed, my comment is mainly about .A attribute! Slicing is separate and 1D dense arrays are probably OK if not ideal IMO.

**Comment by akhmerov** [comment]
> I'm thinking of changing .A to return a sparse array instead, although that is a breaking change. Thoughts?

I hope that PR is a path forward to let the user drop the matrix API altogether. From that perspective I'd say that:

- It's probably best to leave sparse matrix behavior unchanged because the user should be switching away from it anyway
- Not implement `.A` on sparse arrays because the reasoning of #14503 still applies

**Review by dschult — COMMENTED** [review]
Very nice to see this work!  Thanks for this!

I have one minor comment about a test -- and then a question:
Is the long run plan to initially make `csr_array` and friends a subclass of `csr_matrix` and then at some later date rewrite it so that `csr_matrix` is a subclass of `csr_array` and then finally deprecate `csr_matrix`?  Or is that too much long range planning for this stage of the solution?

**Comment by stefanv** [comment]
> * Not implement `.A` on sparse arrays because the reasoning of [Proposal: deprecate spmatrix.A #14503](https://github.com/scipy/scipy/issues/14503) still applies

I've removed the `.A` attribute from the sparse arrays, raising a `NotImplementedError` instead that suggests to use `.todense()`.

**Comment by stefanv** [comment]
I've also removed the `.H` attribute to make it more consistent with ndarrays.  One concern is that taking away these attributes likely breaks the linalg routines for these objects, and that is quite a big cost :/  Perhaps instead of removing them, I should issue deprecation warnings?

**Comment by akhmerov** [comment]
Actually, changing multiplication semantics is likely much more dangerous and will break `sparse.linalg` way worse.

**Comment by akhmerov** [comment]
> True! Fortunately, there is a fairly easy way around that: replace all current instances of * with @.

Indeed. I fear it's not as simple, though, because some of these could stand for scalar multiplication. I brought it up because it seems that fixing `sparse.linalg` is necessary regardless of existence of `.H` and `.A`. Because compatibility with linalg is an issue anyway, I suggest disabling these for sparse arrays after all.

**Comment by stefanv** [comment]
> > True! Fortunately, there is a fairly easy way around that: replace all current instances of * with @.
> 
> Indeed. I fear it's not as simple, though, because some of these could stand for scalar multiplication.

That is the laborious part, yes :)  I will go through each case by hand.

> I brought it up because it seems that fixing `sparse.linalg` is necessary regardless of existence of `.H` and `.A`. Because compatibility with linalg is an issue anyway, I suggest disabling these for sparse arrays after all.

How about I make another PR that moves `scipy.sparse.linalg` over to `@`, then we can deploy this change on top of that?

**Comment by akhmerov** [comment]
> How about I make another PR that moves scipy.sparse.linalg over to `@`, then we can deploy this change on top of that?

I agree, it is indeed likely the best way forward. Also a bit of tooting my own horn, #14496 is ready for review, and it removes some of the matrix API, mainly from tests. It may be useful to merge it before proceeding with this one as well. :wink:

**Comment by rossbar** [comment]
I'm really excited about this proposal - I dipped my toes in a bit to see what it would be like adopting the sparse array interface downstream in networkx/networkx#5139. I noticed a few things from this exercise:

* Reductions along different axes give arrays with different dimensions:

    ```python
    >>> A = scipy.sparse.csr_array([[0, 1, 0, 0],
    ...                             [1, 0, 1, 0],
    ...                             [0, 1, 0, 1],
    ...                             [0, 0, 1, 0]])
    >>> A.sum(axis=0)
    array([[1, 2, 2, 1]])
    >>> A.sum(axis=1)
    array([1, 2, 2, 1])
    ```
  The arrays from the reductions should be 1D to conform to ndarray semantics
    
* There is currently no way to create sparse arrays directly from the various sparse matrix creation functions in `scipy.sparse`. For example, `scipy.sparse.diags` or `scipy.sparse.spdiags` are very commonly used in the NetworkX codebase, and the only way I could figure out how to create a sparse array was to wrap the function like so:

   ```python
   >>> sparr = sp.sparse.csr_array(sp.sparse.spdiags(np.arange(4), 0, 4, 4))
   ```
   
  It might be out-of-scope for this PR, but in terms of adoption this is one of the areas that has required the most change (so far). One idea might be nice to have a separate namespace for these functions that generate sparse object, e.g. something like:
   ```python
   >>> sparr = sp.sparse.array.spdiags(np.arange(4), 0, 4, 4)
   ```
  
* The `getrow` and `getcol` (and for `lil`, `getrowview`) methods still return sparse matrices. I didn't check all the methods for all of the formats systematically, so there may be others.

* Finally, a (partially formed) thought on one consequence of returning dense arrays for indexing. This may lead to "densifying" steps in analyses where the matrix interface wouldn't. I haven't encountered an explicit example, but some situations I saw in the codebase got me thinking along these lines. For example - one common thing from the NX codebase was using `*` with `col` and `row` vectors to compute a sparse outer product. Converting to the array interface naively results in densification:


   ```python
   >>> A.todense()
   array([[0, 1, 0, 0],
          [1, 0, 1, 0],
          [0, 1, 0, 1],
          [0, 0, 1, 0]])
   >>> M = sp.sparse.csr_matrix(A.todense())
   >>> M[:, 0] * M[1, :]  # sparse inputs, sparse output - no dense step
   <4x4 sparse matrix of type '<class 'numpy.int64'>'
          with 2 stored elements in Compressed Sparse Row format>
   >>> A[:, 0][:, np.newaxis] @ A[1, :][np.newaxis, :]
   array([[0, 0, 0, 0],
          [1, 0, 1, 0],
          [0, 0, 0, 0],
          [0, 0, 0, 0]])
   ```
   
   This specific example isn't the most compelling, but at the very least I think this is another thing to watch out for: are there sparse operations with the matrix interface that are no longer possible to replicate (or it's non-obvious how to do so) without densifying when using the array interface?

**Comment by akhmerov** [comment]
@rossbar wouldn't  `A[:, [0]] @ A[[1], :]` address the need to slice without producing dense arrays? (See also my proposal to raise `NotImplementedError` on `A[:, 0]`.)

**Comment by stefanv** [comment]
@rossbar has implemented the change from `*` to `@` in https://github.com/scipy/scipy/pull/14950 !

**Comment by jarrodmillman** [comment]
See #14939.

File: .circleci/config.yml [hunk 1]
```diff
@@ -28,7 +28,7 @@ jobs:
             echo ${CI_PULL_REQUEST//*pull\//} | tee merge.txt
             if [[ $(cat merge.txt) != "" ]]; then
               echo "Merging $(cat merge.txt)";
-              git remote add upstream https://github.com/scipy/scipy.git;
+              git remote add upstream git://github.com/scipy/scipy.git;
```
Comment (rgommers) [inline .circleci/config.yml:31]: spurious change?

May be related to merging master into this branch. Unless you want this PR to be squash-merged, it'd be good to rebase rather than merge master.

File: .gitattributes [hunk 2]
```diff
@@ -3,7 +3,7 @@ tools/win32/build_scripts/nsis_scripts/*.nsi.in eol=crlf
 
 # Don't want to convert line endings of this file, causes test failure on
 # Windows
-scipy/io/_matlab/tests/data/japanese_utf8.txt binary
+scipy/io/matlab/tests/data/japanese_utf8.txt binary
```
Comment (rgommers) [inline .gitattributes:6]: incorrect change as well

File: .circleci/config.yml (hunk 1 above)
Comment (rgommers) [reply .circleci/config.yml:31]: okay, this is re-adding a whole bunch of stuff, can you get rid of the commit and rebase so that the diff only shows sparse changes?

Comment (rgommers) [reply .circleci/config.yml:31]: As @jarrodmillman pointed out, this crossed wires with the PR to make private API use underscores. May need a bit of git fu (or manual redoing of diff) to update in case just removing the merge with master doesn't work.

Comment (stefanv) [reply .circleci/config.yml:31]: Yeah, the problem is that Gagandeep's PR to make things private touched basically single file in scipy.sparse, making it a complete mess to do a rebase (conflicts all along the way).  I therefore took the strategy of cherry-picking his commits and then merging scipy into mine, using the "prefer mine" strategy.  I'm open to tips on how to handle this better.

Comment (rgommers) [reply .circleci/config.yml]: Is there a reason not to squash your commits? There's 9 commits here, and only the first couple are large-ish, the rest are just fix up commits. If you squash them first, you can do a single rebase. 

Alternatively, now that you've done the cherry-picking and merging, just `git reset --mixed <commit-id>` to get a clean diff on top of master, then recommit that.

Comment (stefanv) [reply .circleci/config.yml]: For now, went with squashing all my changes into a single commit and rebasing.

**Comment by stefanv** [comment]
I followed @akhmerov's advice and raised a NotImplementedError on 1D slicing.

There are two follow-up PRs that can be made:

1. Implement a sparse 1D container, so that we can slice out single rows and columns from sparse arrays. These 1D arrays will have to be careful to preserve `dot` semantics.
2. Add sparse array creation utility functions that match those provided for sparse matrices.

**Comment by stefanv** [comment]
@rossbar 

> * Reductions along different axes give arrays with different dimensions:

This is now fixed.

> * The `getrow` and `getcol` (and for `lil`, `getrowview`) methods still return sparse matrices.

It looks like this got fixed along the way; can you still reproduce?

**Comment by ev-br** [comment]
Note to self: remember to change from csr_matrix to csr_array after this is merged and before 1.8.0 (the interpolate dependence on sparse is new in 1.8.0, so can be changed without deprecations) 

https://github.com/scipy/scipy/blob/master/scipy/interpolate/_bspl.pyx#L7

**Comment by rossbar** [comment]
> It looks like this got fixed along the way; can you still reproduce?

I can no longer reproduce on  2d5f36e so this seems fixed:

```python
>>> A = sp.sparse.csr_array([[1,2,0,4],[1,0,0,1],[1,0,2,1],[2,2,1,0.]])
>>> sp_fmts = ["bsr", "coo", "csc", "csr", "dia", "dok", "lil"]
>>> for fmt in sp_fmts:
...     A = A.asformat(fmt)
...     print(type(A), type(A.getrow(0)), type(A.getcol(0)))
<class 'scipy.sparse._arrays.bsr_array'> <class 'scipy.sparse._arrays.csr_array'> <class 'scipy.sparse._arrays.bsr_array'>
<class 'scipy.sparse._arrays.coo_array'> <class 'scipy.sparse._arrays.csr_array'> <class 'scipy.sparse._arrays.csr_array'>
<class 'scipy.sparse._arrays.csc_array'> <class 'scipy.sparse._arrays.csr_array'> <class 'scipy.sparse._arrays.csc_array'>
<class 'scipy.sparse._arrays.csr_array'> <class 'scipy.sparse._arrays.csr_array'> <class 'scipy.sparse._arrays.csr_array'>
<class 'scipy.sparse._arrays.dia_array'> <class 'scipy.sparse._arrays.csr_array'> <class 'scipy.sparse._arrays.csr_array'>
<class 'scipy.sparse._arrays.dok_array'> <class 'scipy.sparse._arrays.csr_array'> <class 'scipy.sparse._arrays.csr_array'>
<class 'scipy.sparse._arrays.lil_array'> <class 'scipy.sparse._arrays.lil_array'> <class 'scipy.sparse._arrays.csr_array'>

**Comment by stefanv** [comment]
:heavy_check_mark: Green!  Okay, folks, let me know what else is needed.

**Comment by rgommers** [comment]
Two questions:
- The test coverage is very small. `sparse.linalg` routines were updated to use `@` instead of `*`, but it's unclear what the status is for this module - for example, does sparse array input give back sparse matrix output still? What are the next steps here, if any?
- Docs: it'd be good to have a summary near the top of http://scipy.github.io/devdocs/reference/sparse.html about the two sets of classes, and a recommendation to users.

File: scipy/sparse/_base.py [hunk 3]
```diff
@@ -557,13 +617,13 @@ def __matmul__(self, other):
...
                              "use '*' instead")
-        return self.__mul__(other)
+        return self._mul_dispatch(other)
 
     def __rmatmul__(self, other):
         if isscalarlike(other):
             raise ValueError("Scalar operands are not allowed, "
                              "use '*' instead")
-        return self.__rmul__(other)
+        return self._rmul_dispatcher(other)
```
Comment (perimosocordiae) [inline scipy/sparse/_base.py:626]: Nitpicking: we have `_mul_dispatch` and `_rmul_dispatcher`. Let's pick one ending and use that for both.

File: scipy/sparse/_base.py [hunk 4]
```diff
@@ -656,26 +716,36 @@ def __pow__(self, other):
...
         elif isscalarlike(other):
             raise ValueError('exponent must be an integer')
         else:
             return NotImplemented
 
     def __getattr__(self, attr):
         if attr == 'A':
+            if self._is_array:
+                warn(np.VisibleDeprecationWarning(
+                    "Please use `.todense()` instead"
```
Comment (perimosocordiae) [inline scipy/sparse/_base.py:739]: I think the plan is to recommend `.toarray()` overall.

File: scipy/sparse/_construct.py [hunk 5]
```diff
@@ -20,10 +20,12 @@
 from ._coo import coo_matrix
 from ._dia import dia_matrix
 
+from ._arrays import coo_array, csr_array, csc_array, dia_array
+
 from ._base import issparse
 
 
-def spdiags(data, diags, m, n, format=None):
+def spdiags(data, diags, m, n, format=None, _array=False):
```
Comment (perimosocordiae) [inline scipy/sparse/_construct.py:28]: I'm not sure how I feel about adding semi-private kwargs to these sparse._construct functions. I would prefer using some tricks with the `format=` kwarg instead, like allowing `format='coo_array'`.

**Review by perimosocordiae — COMMENTED** [review]
Firstly, I want to say thanks! I'm really excited about this PR and the efforts to modernize the scipy.sparse API.

It seems like we could split up the changes here into two pieces:
 - a non-controversial cleanup PR for adding the new "_fmt_container" properties to spmatrix, and all the corresponding changes to that code.
 - the addition of the sparse array API, which I expect we might have more rounds of feedback and review for.

File: scipy/sparse/_construct.py (hunk 5 above)
Comment (akhmerov) [reply scipy/sparse/_construct.py:28]: Perhaps another option here would be a keyword-only argument? Seems like a more convenient interface and more transparent for backwards compatibility at a later stage.

File: scipy/sparse/_base.py (hunk 4 above)
Comment (akhmerov) [reply scipy/sparse/_base.py:739]: `.todense` on a sparse array currently produces a dense array. This is consistent with `.todense` on a sparse matrix producing a dense matrix. Because the name is easier to understand than `.toarray` (huh, don't we have an array already?), I think it makes sense to advertise `.todense` instead of `.toarray`.

**Comment by stefanv** [comment]
> It looks like you didn't yet send a message to the scipy-dev list about this. Could you do that (or link to it in case I missed it)?

Will do.  OK, the new thread is at https://mail.python.org/archives/list/scipy-dev@python.org/thread/TJJ6ACULZVT6CGYGACTI6GLSCZWD7DG2/

**Comment by stefanv** [comment]
> * The test coverage is very small. `sparse.linalg` routines were updated to use `@` instead of `*`, but it's unclear what the status is for this module - for example, does sparse array input give back sparse matrix output still? What are the next steps here, if any?

I have not tested linalg extensively (i.e., not every method) but the few I have tried (e.g., `inv`, `expm`, `eigsh`) behaved as expected (array input gives array output).  Should I parameterize tests to check this for all linalg functionality?

> * Docs: it'd be good to have a summary near the top of http://scipy.github.io/devdocs/reference/sparse.html about the two sets of classes, and a recommendation to users.

I will add that.

**Comment by rgommers** [comment]
> 

> > *
> 
> I have not tested linalg extensively (i.e., not every method) but the few I have tried (e.g., `inv`, `expm`, `eigsh`) behaved as expected (array input gives array output). Should I parameterize tests to check this for all linalg functionality?

Maybe not all tests, that seems excessive regarding runtime - perhaps a single test per function, only with the default parameters?

**Comment by stefanv** [comment]
> Maybe not all tests, that seems excessive regarding runtime - perhaps a single test per function, only with the default parameters?

I've now added tests for all `scipy.sparse.linalg` methods other than `lobpcg`.  My examples were too small for `lobpcg`, so getting it to run without resorting to a dense solver would mean using a much larger input.

File: scipy/sparse/_construct.py (hunk 5 above)
Comment (stefanv) [reply scipy/sparse/_construct.py:28]: This is currently only used to provide `eye` on sparse arrays for `__pow__`, which handles the power zero special case as follows:

```
            if other == 0:
                from ._construct import eye
                return eye(
                    self.shape[0], dtype=self.dtype, _array=self._is_array
                )
```

I can definitely get by without it.

Comment (stefanv) [reply scipy/sparse/_construct.py]: OK, I've removed these.

File: scipy/sparse/_base.py (hunk 3 above)
Comment (stefanv) [reply scipy/sparse/_base.py:626]: Changed.

**Comment by stefanv** [comment]
@perimosocordiae I've addressed the comments you left above.  Do you think this is close enough that we can get it in for 1.8 (slated for Dec 6th)?

File: scipy/sparse/_base.py [hunk 6]
```diff
@@ -656,26 +717,40 @@ def __pow__(self, other):
 
             if other == 0:
                 from ._construct import eye
-                return eye(self.shape[0], dtype=self.dtype)
```
Comment (perimosocordiae) [inline scipy/sparse/_base.py:659]: I'd prefer to keep using `eye` here, and in the `self._is_array` case we can wrap the result with `dia_array` before returning.

File: scipy/sparse/_base.py [hunk 7]
```diff
@@ -1015,27 +1090,29 @@ def sum(self, axis=None, dtype=None, out=None):
...
-                np.ones((n, 1), dtype=res_dtype))
+            ret = self @ self._ascontainer(
+                np.ones((n, 1), dtype=res_dtype)
+            )
 
         if out is not None and out.shape != ret.shape:
             raise ValueError("dimensions do not match")
 
-        return ret.sum(axis=(), dtype=dtype, out=out)
+        return ret.sum(axis=(axis), dtype=dtype, out=out)
```
Comment (perimosocordiae) [inline scipy/sparse/_base.py:1115]: If I'm understanding right, we need to pass `axis` to get a 1-d result for arraylikes, and it'll have no effect for matrix. Is that what's going on here?

We should remove the extra parens around `axis`.

File: scipy/sparse/_index.py [hunk 8]
```diff
@@ -31,18 +31,28 @@ class IndexMixin:
     """
     def __getitem__(self, key):
         row, col = self._validate_indices(key)
+
+        def vector(data):
```
Comment (perimosocordiae) [inline scipy/sparse/_index.py:35]: Rather than a "pass-through or fail" function how about something like this:

```python
def check_1d_slice():
    if self._is_array:
        raise NotImplementedError("...etc etc...")
```

And then we can call it prior to doing any work in the 1d slicing cases.

File: scipy/sparse/_index.py [hunk 9]
```diff
@@ -31,18 +31,28 @@ class IndexMixin:
...
+
         # Dispatch to specialized methods.
         if isinstance(row, INT_TYPES):
             if isinstance(col, INT_TYPES):
                 return self._get_intXint(row, col)
             elif isinstance(col, slice):
-                return self._get_intXslice(row, col)
+                return vector(self._get_intXslice(row, col))
             elif col.ndim == 1:
                 return self._get_intXarray(row, col)
```
Comment (perimosocordiae) [inline scipy/sparse/_index.py:51]: This is also a 1d slice and needs to be checked.

Comment (perimosocordiae) [reply scipy/sparse/_index.py:51]: Also in need of checks: both calls to `_get_arrayXint` below.

File: scipy/sparse/_sputils.py [hunk 10]
```diff
@@ -292,7 +292,7 @@ def check_shape(args, current_shape=None):
     if current_shape is None:
         if len(new_shape) != 2:
             raise ValueError('shape must be a 2-tuple of positive integers')
-        elif new_shape[0] < 0 or new_shape[1] < 0:
+        elif not all(d >= 0 for d in new_shape):
```
Comment (perimosocordiae) [inline scipy/sparse/_sputils.py:295]: elif any(d < 0 for d in new_shape):

File: scipy/sparse/linalg/_dsolve/linsolve.py [hunk 11]
```diff
@@ -222,7 +222,7 @@ def spsolve(A, b, permc_spec=None, use_umfpack=True):
             row_segs = []
             col_segs = []
             for j in range(b.shape[1]):
-                bj = b[:, j].toarray().ravel()
+                bj = b[:, [j]].toarray().ravel()
```
Comment (perimosocordiae) [inline scipy/sparse/linalg/_dsolve/linsolve.py:225]: Leave a TODO comment here to revert this change once sparrays support 1d indexing.

**Review by perimosocordiae — CHANGES_REQUESTED** [review]

**Comment by perimosocordiae** [comment]
I think this is likely to be ready to submit before the 1.8 cutoff.

File: scipy/sparse/linalg/_dsolve/linsolve.py (hunk 11 above)
Comment (stefanv) [reply scipy/sparse/linalg/_dsolve/linsolve.py:225]: I think in this case it makes no difference, since the array is first densified and then raveled.

File: scipy/sparse/_index.py (hunk 9 above)
Comment (stefanv) [reply scipy/sparse/_index.py:51]: Thanks, this helped me catch some bugs---including one that was already there!

Comment (stefanv) [reply scipy/sparse/_index.py:51]: Addressed.

File: scipy/sparse/_index.py (hunk 8 above)
Comment (stefanv) [reply scipy/sparse/_index.py:35]: I don't see how this logic can easily be replaced by that function.  The signature would probably have to be `def check_1d_ix(row, col)`, and then it'd have to duplicate all the logic already in place.  Also, this is probably not the place to be worried about performance, so first doing some work and then bailing seems fine.

Comment (stefanv) [reply scipy/sparse/_index.py:35]: Done.

File: scipy/sparse/_base.py (hunk 7 above)
Comment (stefanv) [reply scipy/sparse/_base.py:1115]: Exactly.

File: scipy/sparse/_base.py (hunk 6 above)
Comment (stefanv) [reply scipy/sparse/_base.py:659]: Done.

**Comment by stefanv** [comment]
Python 3.10 failure unrelated.

File: scipy/sparse/_index.py [hunk 12]
```diff
@@ -149,6 +170,7 @@ def _validate_indices(self, key):
         return row, col
 
     def _asindices(self, idx, length):
+
```
Comment (perimosocordiae) [inline scipy/sparse/_index.py:173]: Remove this blank line?

**Review by perimosocordiae — APPROVED** [review]

File: scipy/sparse/linalg/_dsolve/linsolve.py (hunk 11 above)
Comment (stefanv) [reply scipy/sparse/linalg/_dsolve/linsolve.py:225]: OK, added a note.

**Comment by stefanv** [comment]
Thank you for the thorough reviews, @perimosocordiae!

**Comment by stefanv** [comment]
> Very nice to see this work! Thanks for this!
> 
> I have one minor comment about a test -- and then a question: Is the long run plan to initially make `csr_array` and friends a subclass of `csr_matrix` and then at some later date rewrite it so that `csr_matrix` is a subclass of `csr_array` and then finally deprecate `csr_matrix`? Or is that too much long range planning for this stage of the solution?

@dschult I don't know if I replied to this, it's sometimes a bit hard to see in between all the resolved comments.  But, I think what I was hoping is that we would have `csr_array` as a subclass of `csr_matrix`, then deprecate `csr_matrix`, and then merge the two implementations.

**Comment by stefanv** [comment]
OK, all review comments have been responded to and CI is green for all entries of the matrix that are runnable / pass on main.

**Comment by stefanv** [comment]
@ev-br @tylerjereddy Any chance you'd be willing to give this another look-over? Willing to review for reviews :)

**Comment by ev-br** [comment]
Looks good to me. I think we should give it a drive on 1.8.0, with an understanding that things might need to be tweaked , so a bit of flux is unavoidable. 
An alternative is of course to merge it just after branching 1.8.x, but I'm not sure it's any better.

**Comment by stefanv** [comment]
Having it on 1.8.0 would be great, then we can start working on NetworkX and scikit-learn.

**Comment by jarrodmillman** [comment]
As the release manager of NetworkX, I would like to see this merged before 1.8 as we want to use this API as soon as possible.

**Comment by jarrodmillman** [comment]
@tylerjereddy, @rgommers  Any chance this might be merged in time for 1.8?  We are hoping to start transitioning to the new API in NetworkX as soon as possible:  https://github.com/networkx/networkx/pull/5139

**Comment by larsoner** [comment]
Going for a close-reopen cycle to see if CIs come back green (Azure can't be restarted because it's too old)

**Comment by larsoner** [comment]
We have +1's from @ev-br and @perimosocordiae , it looks like all comments have been addressed (AFAICT), this is something that we really should have, and based on a (light) review of my own this looks good to go. So unless anyone objects or asks for more time, I'll go ahead and merge this in two days!

**Comment by rgommers** [comment]
Agreed, makes sense to merge for 1.8.0 and looks quite good overall. 

Given that there are some missing pieces and limitations still, it'd probably be good to not be aggressive for this first release in pushing users to move to it. Trying it quickly with NetworkX and a few other early adopters seems like a good idea, and once it seems solid, then update the advice for general users.

**Comment by ev-br** [comment]
I guess it needs to be expected that there'll because need to tweak something, so for the 1.8.x series this API is effectively provisional. 
I'd suggest that we craft some wording in the release notes to this effect. Maybe not the p-word directly, but something along the lines of "this is very new, do give it a spin and report back, but do keep an eye on a possible backcompat breakage".

**Review by tylerjereddy — APPROVED** [review]
Ok, CI is green and shows good test coverage.

There is overwhelming support, if not pressure, to put this in, and some folks are keen to make other changes internally & externally as soon as this goes in, so I'll merge it with a bit of time before the cut.

Thanks Stefan et al.

**Comment by tylerjereddy** [comment]
Obviously, this warrants a release notes entry on our wiki

**Comment by WarrenWeckesser** [comment]
Is there any chance someone could fix https://github.com/scipy/scipy/issues/15210 in time for the final version of 1.8.0?

**Comment by arogozhnikov** [comment]
Playing with this new interface (fantastic improvement in experience!).

I don't think this is an intended behavior for conversion to dense:
```
type(coo / 2), type(coo * 2), type(coo / coo.sum(axis=0)), type(coo * coo.sum(axis=0))
```
output:
```
(scipy.sparse._arrays.csc_array,
 scipy.sparse._arrays.csc_array,
 numpy.ndarray,
 scipy.sparse._arrays.coo_array)
```

This may be a very common usecase because of different row-wise / column-wise normalizaitions.


This one may be ok, but I expected this will be optimized:
```
coo[coo > 0.5] = 0
# coo data elements change, but number of elements is the same
```

I also find `.toarray` being confusing, since user already operates with `coo_array`, I think `.todense` should be a recommended way

**Comment by rossbar** [comment]
Great feedback, thanks! In the case of division, I believe the behavior is actually inherited from the sparse matrices:

```python
>>> from scipy.sparse import coo_matrix
>>> coo = coo_matrix(np.arange(100).reshape(10, 10))  # ignoring sparsity for sake of example
>>> type(coo / coo.sum(axis=0))
numpy.matrix
```

> I also find .toarray being confusing, since user already operates with coo_array, I think .todense should be a recommended way

This is also a consequence of consistency with sparse matrices. `spmatrix` defines both a `todense` (which returns a `numpy.matrix` object) and a `toarray` (returns ndarray) method. The sparse arrays provide the same methods, though it is worth noting that `todense` for sparse arrays returns ndarray as one expects:

```python
>>> from scipy.sparse import coo_array
>>> coo = coo_array(np.arange(12).reshape(3, 4))
>>> coo.todense()
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11]])
```

So for new code, I agree that `_sparray.todense` is more clear, but I also think that `_sparray.toarray` around is valuable as it makes it easier to transition interfaces.
